libbeat/processors/add_kubernetes_metadata/matchers.go

```diff
@@ -173,3 +179,122 @@ func (f *FieldFormatMatcher) MetadataIndex(event common.MapStr) string {
 
 	return string(bytes)
 }
+
+// PidMatcher implements matcher that matches container id based on pid from cgroup file
+type PidMatcher struct {
+	matchRegex  string
+	hostPath    string
+	pidField    string
+	ppidField   string
+	pidCidCache *common.Cache
+}
+
+// NewPidMatcher initializes and returns a PidMatcher
+func NewPidMatcher(cfg common.Config) (Matcher, error) {
+	config := struct {
+		RegexField string `config:"matcher_regex"`
+		HostPath   string `config:"host_path"`
+		PidField   string `config:"pid_field"`
+		PpidField  string `config:"ppid_field"`
+	}{}
+
+	err := cfg.Unpack(&config)
+	if err != nil {
+		return nil, fmt.Errorf("fail to unpack the `matcher_regex` configuration: %s", err)
+	}
+
+	if len(config.RegexField) == 0 {
+		// no matcher regex from conf, use default one
+		config.RegexField = "^\\d+:\\w+:\\/.+\\/.+\\/.+\\/([0-9a-f]{64})"
+	}
+
+	if len(config.HostPath) == 0 {
+		// no host path from conf, use default one
+		config.HostPath = "/"
+	}
+
+	if len(config.PidField) == 0 {
+		// no pid field from conf, use default one
+		config.PidField = "process.pid"
+	}
+
+	if len(config.PpidField) == 0 {
+		// no ppid field from conf, use default one
+		config.PpidField = "process.ppid"
+	}
+
+	cache := common.NewCache(time.Minute, 1000)
+	cache.StartJanitor(time.Minute)
+
+	return &PidMatcher{
+		matchRegex:  config.RegexField,
+		hostPath:    config.HostPath,
+		pidField:    config.PidField,
+		ppidField:   config.PpidField,
+		pidCidCache: cache}, nil
+}
+
+// MetadataIndex returns index for matching metadata indexed based on container id
+func (p *PidMatcher) MetadataIndex(event common.MapStr) string {
+	// As we are post processing here, actual process may be already exited.
+	// So trying to find the container id by the parent process id first, hoping parent process lives a bit longer
```

Comment:
Here I changed the order. Ppid is used first. Now i have feeling, that falling back to pid just in case is actually useless, meaning that if cgroup file for ppid is already gone, then child process is closed too, right?

So if i'm not mistaken, then using only ppid for extracting container id should be enough.

Reply:
what if the `ppid` is container runtime itself?

Reply:
right. so in rare cases this pid part is still useful.